app/javascript/components/visualization/ScatterPlot.jsx

```diff
@@ -324,12 +326,15 @@ function RawScatterPlot({
     let [scatter, perfTimes] =
       (clusterResponse ? clusterResponse : [scatterData, null])
     scatter = updateScatterLayout(scatter)
+    const annotIsNumeric = scatter.annotParams.type === 'numeric'
     const layout = scatter.layout
 
     if (filteredCells) {
       const originalData = scatter.data
       const [intersected, plottedIndexes] = intersect(filteredCells, scatter)
-      scatter.data = reassignFilteredCells(plottedIndexes, originalData, intersected)
+      scatter.data = reassignFilteredCells(plottedIndexes, originalData, intersected, annotIsNumeric, setOriginalLabels)
+    } else if (!annotIsNumeric) {
+      setOriginalLabels(getPlottedLabels(scatter.data))
```

Comment:
Would this be be simpler and faster?

```suggestion
      setOriginalLabels(scatter.annotParams.values)
```

I haven't tested that, but it looks like `getPlottedLabels` recomputes on the client what's already available at hand from the server response.

Reply:
Yes, it does - if you have a suggestion for a simpler way to compute the unique entries from a long array, I'm all ears!

Reply:
Agreed, spread/Set seems ideal to uniquify an array, if needed.  Rather, the optimization that crossed my mind was avoiding that need, e.g. maybe using `scatter.annotParams.values` only for cluster-wide annotations, or adjusting color assignment to e.g. always start with red even for study-wide annotations with filtering.  Neither of those seem obviously worthwhile right now, I just note it as food for thought; your approach here is good!